### PR TITLE
Support emitting type-parameter variance in syntax generator

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/66377

### CODE REVIEW

File: src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -1,3 +1,9 @@
+*REMOVED*abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.MethodDeclaration(string name, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.SyntaxNode> parameters = null, System.Collections.Generic.IEnumerable<string> typeParameters = null, Microsoft.CodeAnalysis.SyntaxNode returnType = null, Microsoft.CodeAnalysis.Accessibility accessibility = Microsoft.CodeAnalysis.Accessibility.NotApplicable, Microsoft.CodeAnalysis.Editing.DeclarationModifiers modifiers = default(Microsoft.CodeAnalysis.Editing.DeclarationModifiers), System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.SyntaxNode> statements = null) -> Microsoft.CodeAnalysis.SyntaxNode
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt:1]: these are changes to non-abtract methods.  This is binary-compat safe as this type is not publicly inheritable.

File: src/Workspaces/Core/Portable/Editing/SyntaxGenerator.cs [hunk 2]
```diff
@@ -147,14 +147,29 @@ public SyntaxNode FieldDeclaration(IFieldSymbol field, SyntaxNode? initializer)
         /// <summary>
         /// Creates a method declaration.
         /// </summary>
-        public abstract SyntaxNode MethodDeclaration(
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Editing/SyntaxGenerator.cs:150]: 😕 Why is this showing up now? I see another overload, but it doesn't have optional parameters.

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Editing/SyntaxGenerator.cs:150]: no clue.  i have no idea how this analyzer works. I've *never* found it valuable.  It's far too aggressive and actively gets the rules wrong.  So every time it comes up, it just gets suppressed.

**Review by ryzngard — APPROVED** [review]
LGTM

**Pull request merged**